### PR TITLE
Disables APF if the aggregated apiserver cannot locate the core kube-apiserver

### PR DESCRIPTION
**Pull request by yue9944882**
it's meaningless for an aggregated apiserver to run the APF controller if it cannot even access the core kube-apiserver which serves flowschemas/prioritylevels. 

```release-note
NONE
```

/sig api-machinery
/kind cleanup

/assign @MikeSpreitzer @deads2k @lavalamp

### CODE REVIEW

File: staging/src/k8s.io/apiserver/pkg/server/options/recommended.go [hunk 1]
```diff
@@ -124,12 +125,16 @@ func (o *RecommendedOptions) ApplyTo(config *server.RecommendedConfig) error {
...
-		)
+		if config.ClientConfig != nil {
+			config.FlowControl = utilflowcontrol.New(
+				config.SharedInformerFactory,
+				kubernetes.NewForConfigOrDie(config.ClientConfig).FlowcontrolV1beta1(),
+				config.MaxRequestsInFlight+config.MaxMutatingRequestsInFlight,
+				config.RequestTimeout/4,
+			)
+		} else {
+			klog.Warningf("neither kubeconfig is provided nor service-account is mounted, so APIPriorityAndFairness will be disabled")
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/options/recommended.go:136]: nit "Neither" since it's a message, not an error.

**Comment by deads2k** [comment]
This seems reasonable to me.  I think it's better to run a server than to hard fail in this case.

/approve

I'd like to see the message capitalized.

**Comment by lavalamp** [comment]
I'm not sure, is this really the right thing to do if the main apiserver is just transiently unreachable at startup?

**Comment by yue9944882** [comment]
> I'm not sure, is this really the right thing to do if the main apiserver is just transiently unreachable at startup?

are you thinking about context propagation between the main apiserver and aa-server? at least one fact we're sure is that APF-controller wont be fully working b/c it doesn't know the current api configuration. an alternative to disabling APF controller is to downgrading behavior of APF controller (e.g. make it working w/ static flow-schema and priority-levels)

File: staging/src/k8s.io/apiserver/pkg/server/options/recommended.go (hunk 1 above)
Comment (yue9944882) [reply staging/src/k8s.io/apiserver/pkg/server/options/recommended.go]: fixed

**Comment by MikeSpreitzer** [comment]
You guys recall that the APF filter already has fallback behavior in case no config API objects are available, right?
However, we _did_ make it deliberately stingy.

> I'm not sure, is this really the right thing to do if the main apiserver is just transiently unreachable at startup?

Is this code sensitive to whether a main apiserver is _reachable_?  It looks to me like this code is sensitive to whether this server has configuration for reaching a main apiserver.

**Comment by yue9944882** [comment]
> You guys recall that the APF filter already has fallback behavior in case no config API objects are available, right?
However, we did make it deliberately stingy.

yes, the APF controller assumes an exempt and catch-all priority-level, the latter shares "1" from the global concurrency pool, but given that there're no competitors, i guess the catch-all will take over the global concurrency, is it true?

> that is not a transient matter.

am i missing anything? as far as i know the two thing is transient. if the aa-server is configured to connect the main apiserver, the `ClientConfig` will always be non-nil.

**Comment by MikeSpreitzer** [comment]
Oh, good point about the default config.  The catch-all priority level will get all the concurrency.  So this feature already "fails open" (in the valve sense, not the circuit sense).

I am not sure I understand @yue9944882's comment about transient.  As far as I know, the configuration is _not_ transient, and `config.ClientConfig` will be non-nil whenever this server has configuration for reaching main apiserver(s); whether any main apiserver is reachable at startup is not relevant.

**Comment by MikeSpreitzer** [comment]
BTW, does a custom apiserver even work at all if there is no main apiserver reachable?  What about the TokenReview and SubjectAccessReview requests to the main apiservers?

**Comment by deads2k** [comment]
> I'm not sure, is this really the right thing to do if the main apiserver is just transiently unreachable at startup?

@lavalamp This PR is gated on the presence of a clientconfig.  Without a clientconfig, an apiserver will never make contact with a kube-apiserver.  It's not a transient configuration setting, it is a way for the binary to indicate "I don't connect to a kube-apiserver". Without a connection to a kube-apiserver, I don't see how we would ever get the p&f configuration.

**Comment by lavalamp** [comment]
> Without a clientconfig, an apiserver will never make contact with a kube-apiserver. It's not a transient configuration setting

OK yes I totally misread this, sorry for the noise!

**Pull request merged**